File: src/engine/core/envFile.ts
// Safe read/modify of a .env file (default: .env.local, which is git-ignored).
// Never logs values. Used to persist an API key the user typed during onboarding
// so it lives only in the git-ignored secret file, never in settings or source.

import { existsSync, readFileSync, writeFileSync } from 'node:fs';

import { restrictFile } from './perms.ts';

/**
 * Insert or update `NAME=value` in a .env file, preserving other lines and
 * comments. Creates the file if missing. Returns nothing that reveals the value.
 */
export function upsertEnvVar(path: string, name: string, value: string): void {
  const line = `${name}=${value}`;
  let content = '';
  if (existsSync(path)) content = readFileSync(path, 'utf8');

  const lines = content.length ? content.split(/\r?\n/) : [];
  const idx = lines.findIndex((l) => new RegExp(`^\\s*${escapeRegExp(name)}\\s*=`).test(l));
  if (idx === -1) {
    if (lines.length && lines[lines.length - 1] !== '') lines.push('');
    lines.push(line);
  } else {
    lines[idx] = line;
  }
  // Normalize trailing newline.
  const out = lines.join('\n').replace(/\n*$/, '\n');
  writeFileSync(path, out, { encoding: 'utf8', mode: 0o600 });
  // Also tighten perms on an existing file (mode above only applies on create).
  restrictFile(path);
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

File: src/py/agent_studio/core/env_file.py
"""Safe read/modify of a .env file (default: .env.local, git-ignored).

Mirrors the TS ``upsertEnvVar``: persists an API key the user typed during
onboarding so it lives only in the git-ignored secret file — never in settings
or source. Never logs values; writes owner-only perms.
"""

from __future__ import annotations

import re
from pathlib import Path

from .perms import restrict_file


def upsert_env_var(path: str | Path, name: str, value: str) -> None:
    """Insert or update ``NAME=value``, preserving other lines/comments."""
    path = Path(path)
    content = path.read_text(encoding="utf-8") if path.exists() else ""
    lines = content.split("\n") if content else []
    # Drop a single trailing empty element from a trailing newline.
    if lines and lines[-1] == "":
        lines.pop()

    pattern = re.compile(rf"^\s*{re.escape(name)}\s*=")
    idx = next((i for i, line in enumerate(lines) if pattern.match(line)), -1)
    new_line = f"{name}={value}"
    if idx == -1:
        lines.append(new_line)
    else:
        lines[idx] = new_line

    path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    restrict_file(path)
